src/guest_memory.rs

```diff
@@ -175,19 +196,89 @@ pub trait GuestMemory {
     fn num_regions(&self) -> usize;
 
     /// Return the region containing the specified address or None.
-    fn find_region(&self, GuestAddress) -> Option<&Self::R>;
+    fn find_region(&self, addr: GuestAddress) -> Option<&Self::R>;
 
     /// Perform the specified action on each region.
     /// It only walks children of current region and do not step into sub regions.
-    fn with_regions<F>(&self, cb: F) -> Result<()>
+    fn with_regions<F, E>(&self, cb: F) -> std::result::Result<(), E>
     where
-        F: Fn(usize, &Self::R) -> Result<()>;
+        F: Fn(usize, &Self::R) -> std::result::Result<(), E>;
 
     /// Perform the specified action on each region mutably.
     /// It only walks children of current region and do not step into sub regions.
-    fn with_regions_mut<F>(&self, cb: F) -> Result<()>
+    fn with_regions_mut<F, E>(&self, cb: F) -> std::result::Result<(), E>
     where
-        F: FnMut(usize, &Self::R) -> Result<()>;
+        F: FnMut(usize, &Self::R) -> std::result::Result<(), E>;
+
+    /// Applies two functions, specified as callbacks, on the inner memory regions.
+    ///
+    /// # Arguments
+    /// * `init` - Starting value of the accumulator for the `foldf` function.
+    /// * `mapf` - "Map" function, applied to all the inner memory regions. It returns an array of
+    ///            the same size as the memory regions array, containing the function's results
+    ///            for each region.
+    /// * `foldf` - "Fold" function, applied to the array returned by `mapf`. It acts as an
+    ///             operator, applying itself to the `init` value and to each subsequent elemnent
+    ///             in the array returned by `mapf`.
+    ///
+    /// # Examples
+    ///
+    /// * Compute the total size of all memory mappings in KB by iterating over the memory regions
+    ///   and dividing their sizes to 1024, then summing up the values in an accumulator.
+    ///
+    /// ```
+    /// # #[cfg(feature = "backend-mmap")]
+    /// # fn test_map_fold() -> Result<(), ()> {
+    /// # use vm_memory::{GuestAddress, GuestMemory, GuestMemoryRegion, mmap::GuestMemoryMmap};
+    ///     let start_addr1 = GuestAddress(0x0);
+    ///     let start_addr2 = GuestAddress(0x400);
+    ///     let mem = GuestMemoryMmap::new(&vec![(start_addr1, 1024), (start_addr2, 2048)]).unwrap();
+    ///     let total_size = mem.map_and_fold(
+    ///         0,
+    ///         |(_, region)| region.len() / 1024,
+    ///         |acc, size| acc + size
+    ///     );
+    ///     println!("Total memory size = {} KB", total_size);
+    ///     Ok(())
+    /// # }
+    /// ```
+    fn map_and_fold<F, G, T>(&self, init: T, mapf: F, foldf: G) -> T
```

Comment:
Can't we have those as separate methods aka `map()` and `fold()`? Or not have them at all actually.

I never liked it in Firecracker either and was hoping to get rid of it at some point. We are using it exactly once in Firecracker to handle `KVM_GET_DIRTY_LOG`. It think we can do better.

Reply:
Hmm I wonder if `-> impl Iterator<Item = X>`  would let us implement `map`... I didn't like it very much either, even though I am actually using it here so that makes two...

Reply:
Well, we could provide access to regions via iterators instead of (or in addition to) the `with_regions_*` methods, and then we get `map`, `fold`, and all the other iterator enabled goodies for free :D

Reply:
I tried adding regions_iter but he problem is that iterators have lifetime parameters... I think that would also apply to map() but I am not sure. I think we can improve it later.

Reply:
@bonzini do you mind adding an issue about this one? Just to keep it mind for future refactoring :D